File: Models/Device.swift
import Foundation

struct Device: Codable, Hashable, Identifiable {
    var id: UUID
    var ipAddress: String
    var hostname: String
    var macAddress: String?
    var vendor: String
    var status: DeviceStatus
    var openPorts: [Int]
    var services: [NetworkService]
    var lastSeen: Date
    var isFavorite: Bool

    init(
        id: UUID = UUID(),
        ipAddress: String,
        hostname: String = "",
        macAddress: String? = nil,
        vendor: String = "Unknown",
        status: DeviceStatus = .online,
        openPorts: [Int] = [],
        services: [NetworkService] = [],
        lastSeen: Date = Date(),
        isFavorite: Bool = false
    ) {
        self.id = id
        self.ipAddress = ipAddress
        self.hostname = hostname
        self.macAddress = macAddress
        self.vendor = vendor
        self.status = status
        self.openPorts = openPorts
        self.services = services
        self.lastSeen = lastSeen
        self.isFavorite = isFavorite
    }

    var displayName: String {
        let trimmedHostname = hostname.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmedHostname.isEmpty ? ipAddress : trimmedHostname
    }

    var hasResolvedHostname: Bool {
        let trimmedHostname = hostname.trimmingCharacters(in: .whitespacesAndNewlines)
        return !trimmedHostname.isEmpty && trimmedHostname != ipAddress
    }

    var tableNameSubtitle: String? {
        if hasResolvedHostname {
            return ipAddress
        }

        if vendor != "Unknown" && vendor != "Locally Administered" {
            return vendor
        }

        return macAddress
    }

    var openPortsDisplay: String {
        openPorts.isEmpty ? "-" : openPorts.map(String.init).joined(separator: ", ")
    }

    var servicesDisplay: String {
        services.isEmpty ? "-" : services.map(\.name).joined(separator: ", ")
    }

    var statusSortValue: String {
        status.rawValue
    }

    var nameSortValue: String {
        displayName.localizedLowercase
    }

    var ipSortValue: UInt32 {
        IPv4Address(ipAddress)?.value ?? UInt32.max
    }

    var macSortValue: String {
        macAddress ?? ""
    }

    var vendorSortValue: String {
        vendor.localizedLowercase
    }

    var openPortsSortValue: String {
        openPorts.map { String(format: "%05d", $0) }.joined(separator: ",")
    }

    var servicesSortValue: String {
        servicesDisplay.localizedLowercase
    }

    var hasWebService: Bool {
        openPorts.contains(80) || openPorts.contains(443) || openPorts.contains(8080)
    }

    var hasSSH: Bool {
        openPorts.contains(22)
    }

    var hasSMB: Bool {
        openPorts.contains(445)
    }

    var hasVNC: Bool {
        openPorts.contains(5900)
    }

    func matches(_ other: Device) -> Bool {
        if let macAddress, let otherMAC = other.macAddress, macAddress.caseInsensitiveCompare(otherMAC) == .orderedSame {
            return true
        }
        return ipAddress == other.ipAddress
    }
}

File: Windows/LanScope.Core/Models/Device.cs
using System.ComponentModel;
using System.Runtime.CompilerServices;
using System.Text.Json.Serialization;

namespace LanScope.Core.Models;

public sealed class Device : INotifyPropertyChanged
{
    private bool _isFavorite;

    public Guid Id { get; set; } = Guid.NewGuid();
    public string IpAddress { get; set; } = "";
    public string Hostname { get; set; } = "";
    public string? MacAddress { get; set; }
    public string Vendor { get; set; } = "Unknown";
    public DeviceStatus Status { get; set; } = DeviceStatus.Online;
    public List<int> OpenPorts { get; set; } = [];
    public List<NetworkService> Services { get; set; } = [];
    public DateTimeOffset LastSeen { get; set; } = DateTimeOffset.Now;

    public bool IsFavorite
    {
        get => _isFavorite;
        set
        {
            if (_isFavorite == value) return;
            _isFavorite = value;
            OnPropertyChanged();
            OnPropertyChanged(nameof(FavoriteGlyph));
        }
    }

    [JsonIgnore] public string DisplayName => string.IsNullOrWhiteSpace(Hostname) ? IpAddress : Hostname;
    [JsonIgnore] public string OpenPortsDisplay => OpenPorts.Count == 0 ? "-" : string.Join(", ", OpenPorts);
    [JsonIgnore] public string ServicesDisplay => Services.Count == 0 ? "-" : string.Join(", ", Services.Select(x => x.Name));
    [JsonIgnore] public string StatusDisplay => Status.ToString();
    [JsonIgnore] public string FavoriteGlyph => IsFavorite ? "★" : "▣";
    [JsonIgnore] public bool HasWebService => OpenPorts.Any(x => x is 80 or 443 or 8080);
    [JsonIgnore] public bool HasSsh => OpenPorts.Contains(22);
    [JsonIgnore] public bool HasSmb => OpenPorts.Contains(445);
    [JsonIgnore] public bool HasRdp => OpenPorts.Contains(3389);

    public bool Matches(Device other) =>
        (!string.IsNullOrWhiteSpace(MacAddress) &&
         !string.IsNullOrWhiteSpace(other.MacAddress) &&
         string.Equals(MacAddress, other.MacAddress, StringComparison.OrdinalIgnoreCase)) ||
        string.Equals(IpAddress, other.IpAddress, StringComparison.OrdinalIgnoreCase);

    public event PropertyChangedEventHandler? PropertyChanged;
    private void OnPropertyChanged([CallerMemberName] string? name = null) =>
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
}

public enum DeviceStatus
{
    Online,
    Offline,
    Unknown
}

public sealed record NetworkService(int Port, string Name, string? Scheme = null)
{
    [JsonIgnore] public string DisplayName => $"{Name} :{Port}";
}
